### PR TITLE
[SYCL] Add support for templated call operator in functors

### PR DESCRIPTION
**Pull request by srividya-sundaram**
This PR enables kernels to be defined as named function object types, where the  operator() member function is templated.

Example:
```
class FunctorWithCallOpTemplated {
  int x;
  public:
  template <int x = 0>
  void operator()() const {}
};

q.submit([&](sycl::handler &cgh) {
    cgh.single_task(FunctorWithCallOpTemplated{});
  });
```

### CODE REVIEW

File: clang/test/SemaSYCL/undefined-functor.cpp [hunk 1]
```diff
@@ -18,6 +18,13 @@ class FunctorWithCallOpDefined {
   void operator()() const {}
 };
 
+class FunctorWithCallOpTemplated {
+  int x;
```
Comment (premanandrao) [inline clang/test/SemaSYCL/undefined-functor.cpp:22]:
```suggestion
```

**Review by premanandrao — APPROVED** [review]
LGTM

**Comment by srividya-sundaram** [comment]
@intel/llvm-gatekeepers Hi, this is ready for merge. Thank you.

**Review by smanna12 — APPROVED** [review]

**Comment by srividya-sundaram** [comment]
Hi @tfzhu @yinyangsx 
The `SYCL / Linux / Build + LIT` job has been waiting forever to be picked up from the queue. I haven't seen them take this long.
Can you please take a look? Thank you!

**Comment by tfzhu** [comment]
> Hi @tfzhu @yinyangsx The `SYCL / Linux / Build + LIT` job has been waiting forever to be picked up from the queue. I haven't seen them take this long. Can you please take a look? Thank you!

2 Linux runners are offline. Fixed and task should be started later

**Comment by srividya-sundaram** [comment]
@tfzhu I can't tell what's wrong with `SYCL / Linux / [AWS] CUDA LLVM Test Suite (pull_request_target)` by looking at the log.
Can you please comment if you are aware of what's going on?

**Pull request merged**